vllm/spec_decode/ngram_worker.py

```diff
@@ -0,0 +1,260 @@
+import copy
+from typing import Dict, List, Optional
+
+import numpy as np
+import torch
+
+from vllm.sequence import SamplerOutput, SequenceGroupMetadata
+from vllm.spec_decode.interfaces import SpeculativeProposals
+from vllm.spec_decode.top1_proposer import DraftModelTop1Proposer
+from vllm.worker.worker import Worker
+
+
+class NGramWorker(Worker):
+    """NGramWorker provides a light drafter without need for model.
+
+    Current NGramWorker only implement prompt lookup decoding,
+    and in future we may also do RAG type drafter and other scenerios
+    which don't rely on LLM model to give proposals.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Lazy initialization list.
+        self._proposer: DraftModelTop1Proposer
+
+    def fake_load_model(self):
+        pass
+
+    def fake_execute_model(
+        self,
+        seq_group_metadata_list: Optional[List[SequenceGroupMetadata]] = None,
+        blocks_to_swap_in: Optional[Dict[int, int]] = None,
+        blocks_to_swap_out: Optional[Dict[int, int]] = None,
+        blocks_to_copy: Optional[Dict[int, List[int]]] = None,
+        num_lookahead_slots: int = 0,
+    ):
+        pass
+
+    def set_ngram_window_size(self, ngram_prompt_lookup_max: int,
+                              ngram_prompt_lookup_min: int):
+        self.ngram_prompt_lookup_max = ngram_prompt_lookup_max
+        self.ngram_prompt_lookup_min = ngram_prompt_lookup_min
+
+    def init_device(self):
+        super().init_device()
+        # Since ngram worker use no model, fake load_model/execute_model
+        self.load_model = self.fake_load_model
+        self.execute_model = self.fake_execute_model
+
+        self._proposer = DraftModelTop1Proposer(
+            self,
+            self.device,
+            self.max_model_len,
+            self.vocab_size,
+        )
+
+    @torch.inference_mode()
+    def execute_model_multi_step(
+        self,
+        seq_group_metadata_list: List[SequenceGroupMetadata],
+        blocks_to_swap_in: Dict[int, int],
+        blocks_to_swap_out: Dict[int, int],
+        blocks_to_copy: Dict[int, List[int]],
+        num_steps: int,
+    ) -> Optional[List[SamplerOutput]]:
+        """Run the model forward pass num_steps times. Returns the list of
+        sampler output, one per model forward pass.
+        """
+        self._raise_if_unsupported(seq_group_metadata_list, blocks_to_swap_in,
+                                   blocks_to_swap_out, blocks_to_copy)
+
+        # Shallow copy input data so modifications (such as appending tokens)
+        # do not cause side-effects.
+        copied_seq_group_metadata_list = self._shallow_copy_inputs(
+            seq_group_metadata_list)
```

Comment:
IIUC, appending speculative tokens is a general step for every speculative decoding algorithms. @cadedaniel could you help confirm whether we need to have this shallow copy or not? Thanks.

Reply:
yeah, we should remove all this multi step stuff for ngram

Reply:
I believe we don't need modification to seq_group_metadata_list, which is the only motivation for the shallow copy. Can we live without it in the ngram worker?

Reply:
Would apply it in latest commit~